### ISSUE TITLE
Too Many Arguments Error when st2 run terraform.init plan_path

### ISSUE THREAD

**Issue by setswei** [StackStorm-Exchange/stackstorm-terraform#41]
Hello, 

I am receiving the following error when I try to use the terraform.init action in stack storm

```
id: 63be204d1ffffbbf817002b4
action.ref: terraform.init
context.user: st2admin
parameters: 
  plan_path: /opt/stackstorm/packs/DIRECTORY/
status: failed
start_timestamp: Wed, 11 Jan 2023 02:34:53 UTC
end_timestamp: Wed, 11 Jan 2023 02:34:54 UTC
result: 
  exit_code: 0
  result: null
  stderr: 'Too many command line arguments. Did you mean to use -chdir?
    '
  stdout: ''
```

The /opt/stackstorm/packs/DIRECTORY/ has all of my terraform templates in it and the action step itself is basic

```
  task2:
    action: terraform.init
    input:
      plan_path: /opt/stackstorm/packs/DIRECTORY/
      terraform_exec: terraform
```

I have confirmed that terraform is installed and all python dependencies are as well. Has anyone else experienced this error or am I missing something?

Thanks

**Comment by martezr** [comment]
I experienced this same issue a few weeks back but thought it was something related to my environment. It looks likes the migration to the new library is the culprit #36 but it seems like this issue would have existed for several months in that case. The fix should be removing the code in the actions to change directory `os.chdir(plan_path)` and using the `workding_dir` named variable for the plan_path instead of just passing it to the function.

The following snippet is an example of the changes for the init action.

```
import os
from lib import action
from python_terraform import IsFlagged, IsNotFlagged


class Init(action.TerraformBaseAction):
    def run(self, plan_path, terraform_exec, backend, upgrade):
        """
        Initialize a working directory containing Terraform configuration files

        Args:
        - plan_path: path of the Terraform files
        - terraform_exec: path of the Terraform bin
        - backend: backend configuration variable file
        - upgrade: Run init with -upgrade option

        Returns:
        - dict: Terraform init command output
        """
        self.terraform.terraform_bin_path = terraform_exec
        self.terraform.working_dir = plan_path

        return_code, stdout, stderr = self.terraform.init(
            backend_config=backend,
            capture_output=False,
            upgrade=IsFlagged if upgrade else IsNotFlagged
        )

        return self.check_result(return_code, stdout, stderr)
```

**Comment by setswei** [comment]
Hey @martezr,

Thank you for your response. I can confirm that your code snippet has resolved the issue with the init.py file. Subsequently, the same issue occurs on all stackstorm actions in the pack.

I have modified the plan.py and apply.py files with the code snippets below and these now work.

plan.py
```
import os
from lib import action


class Plan(action.TerraformBaseAction):
    def run(self, plan_path, state_file_path, target_resources, terraform_exec,
            variable_dict, variable_files):
        """
        Plan the changes required to reach the desired state of the configuration

        Args:
        - plan_path: path of the Terraform files
        - state_file_path: path of the Terraform state file
        - target_resources: list of resources to target from the configuration
        - terraform_exec: path of the Terraform bin
        - variable_dict: dictionary of Terraform variables that will overwrite the
            variable files if both are declared
        - variable_files: array of Terraform variable files

        Returns:
        - dict: Terraform output command output
        """

        self.terraform.working_dir = plan_path
        self.terraform.state = state_file_path
        self.terraform.targets = target_resources
        self.terraform.terraform_bin_path = terraform_exec
        self.terraform.var_file = variable_files
        self.terraform.variables = variable_dict

        return_code, stdout, stderr = self.terraform.plan(capture_output=False)

        return self.check_result(
            return_code,
            stdout,
            stderr,
            return_output=True,
            valid_return_codes=[0, 2]
        )
``` 

apply.py
```
import os
from lib import action
from python_terraform import IsFlagged


class Apply(action.TerraformBaseAction):
    def run(self, plan_path, state_file_path, target_resources, terraform_exec,
            variable_dict, variable_files):
        """
        Apply the changes required to reach the desired state of the configuration.

        Args:
        - plan_path: path of the Terraform files
        - state_file_path: path of the Terraform state file
        - target_resources: list of resources to target from the configuration
        - terraform_exec: path of the Terraform bin
        - variable_dict: dictionary of Terraform variables that will overwrite the
            variable files if both are declared
        - variable_files: array of Terraform variable files

        Returns:
        - dict: Terraform output command output
        """

        self.terraform.working_dir = plan_path
        self.terraform.state = state_file_path
        self.terraform.targets = target_resources
        self.terraform.terraform_bin_path = terraform_exec
        self.terraform.var_file = variable_files
        self.terraform.variables = variable_dict

        return_code, stdout, stderr = self.terraform.apply(
            skip_plan=True,
            auto_approve=IsFlagged,
            capture_output=False
        )

        return self.check_result(return_code, stdout, stderr, return_output=True)
```

Would this be considered the permanent fix then? or would it need to be investigated further.

**Comment by martezr** [comment]
Yes, this would be the permanent fix for this issue. A pull request with the changes and associated pack updates (changelog, tests, etc.) just needs to be opened for review.

**Comment by setswei** [comment]
Hey @martezr .

I would happy to open the PR I have found another issue and already found a fix so I will add that one too. I will submit it in the next day or 2.

Thanks for pointing me in the right direction.

**Comment by setswei** [comment]
Hey @martezr,

I have created the pull requests to address the issues discussed above. but I am struggling to get the tests to work (I have not had any experience with the mock testing library. 

Tests are failing for actions I didn't even touch, have these tests been broken for a while?

I am currently reading up on the mock library and will try to fix it. but any assistance to speed this up would be appreciated.

**Comment by martezr** [comment]
It looks like a number of the errors are unresolved from previous commits to the repository. 

An issue that's cause several failures is how the pack has been updated to locate the Terraform version that the underlying library requires. This is because the CI tests doesn't include a step to pull down the Terraform binary or mock the value the call is supposed to make.

There are mocks in the test that refer to `mock_chdir` which is no longer necessary for the actions that have been migrated to using the native chdir functionality of the underlying library.

Additionally, it looks like all the tests weren't updated after the migration to the new library was made and there is a bit of rabbit hole with errors.

I might get an opportunity this week to dig through updating all the tests for the pack to get the CI to go green.

**Comment by setswei** [comment]
issue has been fixed in PR #44
